File: src/models/Message.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IMessage extends Document {
    from: mongoose.Schema.Types.ObjectId;
    channel: mongoose.Schema.Types.ObjectId;
    content: string; // Encrypted content
    timestamp: Date;
    status: 'sent' | 'delivered' | 'read';
    edited: boolean;
    editedAt?: Date;
    attachments: string[];
    messageType: 'text' | 'image' | 'file' | 'system';
}

const MessageSchema: Schema = new Schema({
    from: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    channel: { type: mongoose.Schema.Types.ObjectId, ref: 'Channel', required: true },
    content: { type: String, required: true },
    timestamp: { type: Date, default: Date.now },
    status: { type: String, enum: ['sent', 'delivered', 'read'], default: 'sent' },
    edited: { type: Boolean, default: false },
    editedAt: { type: Date },
    attachments: [{ type: String }],
    messageType: { type: String, enum: ['text', 'image', 'file', 'system'], default: 'text' },
});

export default mongoose.model<IMessage>('Message', MessageSchema);

File: Driftway-Microservices/text-channels/lib/text_channels/messaging/message.ex
defmodule TextChannels.Messaging.Message do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  schema "messages" do
    field :content, :string
    field :encrypted_content, :string
    field :channel_id, :binary_id
    field :user_id, :binary_id
    field :username, :string
    field :message_type, :string, default: "text"
    field :status, :string, default: "sent"
    field :edited, :boolean, default: false
    field :reply_to, :binary_id
    field :reactions, {:array, :map}, default: []
    field :attachments, {:array, :map}, default: []

    timestamps()
  end

  def changeset(message, attrs) do
    message
    |> cast(attrs, [:content, :encrypted_content, :channel_id, :user_id, :username, 
                    :message_type, :status, :edited, :reply_to, :reactions, :attachments])
    |> validate_required([:channel_id, :user_id, :username])
    |> validate_length(:content, max: 2000)
    |> validate_inclusion(:message_type, ["text", "image", "file", "system"])
    |> validate_inclusion(:status, ["sent", "delivered", "read", "failed"])
  end
end